pkg/store/copr/coprocessor.go

```diff
@@ -190,6 +190,17 @@ func (c *CopClient) BuildCopIterator(ctx context.Context, req *kv.Request, vars
 		buildTaskElapsed: *buildOpt.elapsed,
 		runawayChecker:   req.RunawayChecker,
 	}
+	// Pipelined-dml can flush locks when it is still reading.
+	// The coprocessor of the txn should not be blocked by itself.
+	// It should be the only case where a coprocessor can read locks of the same ts.
+	//
+	// But when start_ts is not obtained from PD,
+	// the start_ts could conflict with another pipelined-txn's start_ts.
+	// in which case the locks of same ts cannot be ignored.
+	// We rely on the assumption: start_ts is not from PD => this is a stale read.
+	if !req.IsStaleness {
```

Comment:
For `tidb_snapshot` read path, the `staleness` may not be set.

Reply:
This is a bit confusing. Is it intentional or an oversight? I suppose that `tidb_snapshot` is solely used for stale reads.

Reply:
The `stale` flag is not set in the kv request if I do not remember wrong, there are two stale read implementaions in tidb and `tidb_snasphot` is the earlier one.

Reply:
Yes I've verified that. I mean if it's unintentional we can fix it by setting `IsStaleness` for this case.

Reply:
Do you mean setting `IsStaleness` for `tidb_snaoshot` read?

Reply:
Yes

Reply:
In theory, this is the way to proceed with the change. 
However, it's a breaking change because `tidb_snapshot` wasn't set before (setting it will trigger a check on `safe_ts`). Currently, it's unclear if other components rely on this behavior. 
We can proceed by merging this PR first.

Reply:
Then it's possible that when `tidb_snapshot` is set to a ts conflicting with a pipelined txn, coprocessor ignores the locks while point_get doens't?

Reply:
I think so though `tidb_snapshot` seems not to be used widely currenly.